### PR TITLE
Slack channel

### PR DESCRIPTION
**Pull request by rasoro**
https://github.com/rapidpro/rapidpro/issues/1661

### CODE REVIEW

File: handlers/slack/slack.go [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+
+// mtPayload is a struct that represents the body of a SendMmsg text part
+type mtPayload struct {
+	Channel string `json:"channel"`
+	Text    string `json:"text"`
+}
+
+// moPayload is a struct that represents message payload from message type event
+type moPayload struct {
+	Token    string `json:"token,omitempty"`
```
[the remark this answers is not in the document]
Comment (rowanseymour) [reply handlers/slack/slack.go:349]: Maybe useful to add a url in the comment for structs like this that points to the API docs for that thing

File: handlers/slack/slack.go [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+
+	// if event is not a message or is from the bot ignore it
+	if strings.Contains(payload.Event.Type, "message") && payload.Event.BotID == "" {
+
+		date := time.Unix(int64(payload.EventTime), 0)
+
+		var userName string
+		var path string
+		if payload.Event.ChannelType == "channel" { //if is a message from a slack channel that bot is in
+			path = payload.Event.Channel
```
[the remark this answers is not in the document]
Comment (rasoro) [reply handlers/slack/slack.go:84]: maybe we could have both, I think maybe we can route contacts in the channel by their `userID`  with a URN like `slack:channelID:userID`. When we need to specify the contact we use the `userID`, when we need to specify the sending channel we use the `channelID`.

what do you think about this?

I think I would have to make some adjustments to gocommon to get it that way.

File: handlers/slack/slack.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rasoro) [reply handlers/slack/slack.go:349]: > On all the types here can only have the fields we'll use only

sure, i'll clean that, thanks for the observation.



> Maybe useful to add a url in the comment for structs like this that points to the API docs for that thing

I'll arrange that in the next commits, thanks.

File: handlers/slack/slack.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (norkans7) [reply handlers/slack/slack.go:84]: I do not think we have a good way to support a flow to interact with a group of contacts(such as slack channels, telegram groups, WhatsApp groups)
the challenges are when will the flow advance to the next node? when the first contact reply or when all the contacts in the group replied?
I really think we need to keep supporting the 1 on 1 message interactions only

File: handlers/slack/slack.go [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+	// if event is not a message or is from the bot ignore it
+	if payload.Event.Type == "message" && payload.Event.BotID == "" && payload.Event.ChannelType == "im" {
+
+		date := time.Unix(int64(payload.EventTime), 0)
+
+		userInfo, log, err := getUserInfo(payload.Event.User, channel)
+		if err != nil {
+			h.Backend().WriteChannelLogs(ctx, []*courier.ChannelLog{log})
+			return nil, handlers.WriteAndLogRequestError(ctx, h, channel, w, r, err)
+		}
```
Comment (norkans7) [inline handlers/slack/slack.go:85]: These changes seem to belong to a method such as `DescribeURN` and that will make sure we are honouring the workspace isAnon status

**Review by norkans7 — APPROVED** [review]

**Review by rowanseymour — APPROVED** [review]

**Pull request merged**